src/addons/meta/runtime.c

```diff
@@ -0,0 +1,328 @@
+    if (ctor_hook_required) {
+      ecs_meta_rtt_call_data_t *ctor_data = ecs_vec_append_t(NULL, &rtt_ctx->ctors, ecs_meta_rtt_call_data_t);
+      ctor_data->count = m->count;
+      ctor_data->offset = m->offset;
+      ctor_data->type_info = member_ti;
+      if (member_ti->hooks.ctor) {
+        ctor_data->hook.xtor = member_ti->hooks.ctor;
+      } else {
+        ctor_data->hook.xtor = flecs_default_ctor;
+      }
+    }
+    if (dtor_hook_required && member_ti->hooks.dtor) {
+      ecs_meta_rtt_call_data_t *dtor_data = ecs_vec_append_t(NULL, &rtt_ctx->dtors, ecs_meta_rtt_call_data_t);
+      dtor_data->count = m->count;
+      dtor_data->offset = m->offset;
+      dtor_data->type_info = member_ti;
+      dtor_data->hook.xtor = member_ti->hooks.dtor;
+    }
+    if (move_hook_required) {
+      ecs_meta_rtt_call_data_t *move_data = ecs_vec_append_t(NULL, &rtt_ctx->moves, ecs_meta_rtt_call_data_t);
+      move_data->offset = m->offset;
+      move_data->type_info = member_ti;
+      if (member_ti->hooks.move) {
+        move_data->count = m->count;
+        move_data->hook.move = member_ti->hooks.move;
+      } else {
+        move_data->count = m->size * m->count;
+        move_data->hook.move = flecs_meta_rtt_default_move;
+      }
+    }
+    if (copy_hook_required) {
+      ecs_meta_rtt_call_data_t *copy_data = ecs_vec_append_t(NULL, &rtt_ctx->copys, ecs_meta_rtt_call_data_t);
+      copy_data->offset = m->offset;
+      copy_data->type_info = member_ti;
+      if (member_ti->hooks.copy) {
+        copy_data->count = m->count;
+        copy_data->hook.copy = member_ti->hooks.copy;
+      } else {
+        copy_data->count = m->size * m->count;
+        copy_data->hook.copy = flecs_meta_rtt_default_copy;
+      }
+    }
+  }
+}
+
+// Checks if an array's underlying type has hooks installed. If so, it generates and installs required hooks for the
+// array type itself. These hooks will invoke the underlying type's hook for each element in the array.
+static void flecs_meta_rtt_init_default_hooks_array(ecs_world_t *world, ecs_entity_t component,
+                                                    const ecs_type_info_t *ti) {
+  const EcsArray *array_info = ecs_get(world, component, EcsArray);
+  ecs_assert(array_info != NULL, ECS_INTERNAL_ERROR, NULL);
+  const ecs_type_info_t *array_ti = ecs_get_type_info(world, array_info->type);
+  bool ctor_hook_required = array_ti->hooks.ctor && array_ti->hooks.ctor != flecs_default_ctor;
+  bool dtor_hook_required = array_ti->hooks.dtor != NULL;
+  bool move_hook_required = array_ti->hooks.move != NULL;
+  bool copy_hook_required = array_ti->hooks.copy != NULL;
+
+  ecs_meta_rtt_ctx_t *rtt_ctx = flecs_meta_rtt_configure_hooks(world, ti, ctor_hook_required, dtor_hook_required,
+                                                               move_hook_required, copy_hook_required);
+
+  if (!rtt_ctx)
+    return;  // no hooks required
+
+  if (ctor_hook_required) {
```

Comment:
I am not sure I understand what you mean. By the first if-else-if statement in your suggestion you're implying that there will be either ctor or dtor but not both.
Also, each hook must have its own entry. Even if they both have the same type info, offset and count, I am sharing the context type `ecs_meta_rtt_ctx_t` and call data information `ecs_meta_rtt_call_data_t` type for both structs and arrays.
In a future iteration, we could have a `ecs_meta_rtt_ctx_t` that is specific for arrays, as an optimization. This would have a common count/offset/type_info at the context level, saving memory and make that code shorter as you want.

Reply:
My bad, I've updated my example to fix the issue. It's just a suggestion to reduce the amount of code duplication.

Reply:
Still, I am sharing `ecs_meta_rtt_ctx_t` for both structs and arrays and I need to add an entry for each hook type, so unfortunately I have to initialize each `ecs_meta_rtt_call_data_t`.
Anyway, I removed as much of the redundancy. Take a look now and see if that works.

Reply:
You're right, LGTM